src/widgets/modals/DataTablePageModal.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * mSupply Mobile
+ * Sustainable Solutions (NZ) Ltd. 2019
+ */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import ModalContainer from './ModalContainer';
+import { getModalTitle, MODAL_KEYS } from '../../utilities/getModalTitle';
+import { AutocompleteSelector } from '../AutocompleteSelector';
+import { TextEditor } from '../TextEditor';
+import { UIDatabase } from '../../database';
+import { modalStrings } from '../../localization';
+
+/**
+ * Wrapper around ModalContainer, containing common modals used in various
+ * DataTable pages.
+ * @prop {Bool}   fullScreen   Force the modal to cover the entire screen.
+ * @prop {Bool}   isOpen       Whether the modal is open
+ * @prop {Func}   onClose      A function to call if the close x is pressed
+ * @prop {Func}   onSelect     The components to render within the modal
+ * @prop {String} modalKey     The title to show in within the modal.
+ * @prop {String} currentValue The current value a modal should be i.e. theirRef/comment
+ */
+export const DataTablePageModal = ({
+  fullScreen,
+  isOpen,
+  onClose,
+  modalKey,
+  onSelect,
+  currentValue,
+}) => {
+  const ModalContent = () => {
+    switch (modalKey) {
+      case MODAL_KEYS.ITEM_SELECT:
+        return (
+          <AutocompleteSelector
+            options={UIDatabase.objects('Item')}
+            queryString="name BEGINSWITH[c] $0 OR code BEGINSWITH[c] $0"
+            queryStringSecondary="name CONTAINS[c] $0"
+            sortByString="name"
+            onSelect={onSelect}
+            renderLeftText={item => `${item.name}`}
+            renderRightText={item => `${item.totalQuantity}`}
+          />
+        );
+
+      case MODAL_KEYS.COMMENT_EDIT:
+        return <TextEditor text={currentValue} onEndEditing={onSelect} />;
+
+      case MODAL_KEYS.THEIR_REF_EDIT:
+        return <TextEditor text={currentValue} onEndEditing={onSelect} />;
+
+      case MODAL_KEYS.SELECT_CUSTOMER:
+        return (
+          <AutocompleteSelector
+            options={UIDatabase.objects('Customer')}
+            isOpen={isOpen}
+            placeholderText={modalStrings.start_typing_to_select_customer}
+            queryString="name BEGINSWITH[c] $0"
+            sortByString="name"
+            onSelect={onSelect}
+          />
+        );
+
+      case MODAL_KEYS.SELECT_SUPPLIER:
```

Comment:
I think it's worth thinking about the direction here, soon every modal throughout the app's pages will be in this one file. As we continue with this approach eventually the various pages will be virtually redundant and to get an idea of what any one page looks like requires looking at several files.

Thoughts?

Reply:
I think you will have to look at several files to fully understand a page. But I think that's ok. Essentially very similar/exact code is duplicated in 8+ pages. I think this consolidation will reduce bugs/inconsistencies and actually make it easier, after a small learning curve - you understand how this component works - it will be easier to understand the next page a little easier. Changes also become easier - one place to make a change effecting a large chunk/all of the app.

Reply:
Fair, just apprehension from mSupply desktop; done adhoc and with process variables the same idea is a nightmare.

Reply:
Lets see how it pans out 😄 currently it's fine though.